### PR TITLE
Added `time_attr` to `LinkNeighborLoader`

### PR DESCRIPTION
**Pull request by wsad1**
This PR make the following changes.
1. Adds `time_attr` to `LinkNeighborLoader`
2. Refactors `NeigbhorSampler` and `LinkNeighborSampler`. Common code is moved to `_sparse_neighbor_sampler` and `hetero_sparse_neighbor_sampler`.

### CODE REVIEW

File: torch_geometric/loader/neighbor_loader.py [hunk 1]
```diff
@@ -186,60 +186,67 @@ class _DataArgument(object):
...
+            self.colptr,
+            self.row,
+            index,
+            self.num_neighbors,
+            self.replace,
+            self.directed,
+        )
+        return node, row, col, edge
+
+    def _hetero_sparse_neighbor_sample(self, index_dict):
```
Comment (rusty1s) [inline torch_geometric/loader/neighbor_loader.py:201]:
```suggestion
    def _hetero_sparse_neighbor_sample(self, index_dict: Dict[str, Tensor]):
```

**Review by rusty1s — APPROVED** [review]
This is really cool and clean. Thank you!

File: torch_geometric/loader/link_neighbor_loader.py [hunk 2]
```diff
@@ -216,6 +192,11 @@ class LinkNeighborLoader(torch.utils.data.DataLoader):
             :meth:`F.binary_cross_entropy`) and of type
             :obj:`torch.long` for multi-class classification (to facilitate the
             ease-of-use of :meth:`F.cross_entropy`). (default: :obj:`0.0`).
+        time_attr (str, optional): The name of the attribute that denotes
```
[the remark this answers is not in the document]
Comment (wsad1) [reply torch_geometric/loader/link_neighbor_loader.py:199]: updated the docs.

**Comment by Padarn** [comment]
This is nice. Thanks!

**Pull request merged**